docs/proposals/20220209-machinepool-machines.md

```diff
@@ -0,0 +1,195 @@
+* **MachinePool**: a spec for an infrastructure component that can scale a group of Machines. A MachinePool acts similarly to a [MachineDeployment][], but leverages scaling up resources that are specific to an infrastructure provider. However, while individual Machines in a MachineDeployment are visible and can be removed, individual Machines in a MachinePool are not represented.
+* **MachinePool Machine**: a Machine that is owned by a MachinePool. It requires no special treatment in Cluster API in general; a Machine is a Machine without regard to ownership.
+
+See also the [Cluster API Book Glossary][].
+
+## Summary
+
+MachinePools should be enhanced to own Machines that represent each of its `Replicas`.
+
+These "MachinePool Machines" will open up the following opportunities:
+
+- Share common behavior for Machines in MachinePools, which otherwise would need to be implemented by each infrastructure provider.
+- Allow both Cluster API-driven and externally driven interactions with individual Machines in a MachinePool.
+- Enable MachinePools to use deployment strategies similarly to MachineDeployments.
+
+## Motivation
+
+MachineDeployments and Machines provide solutions for common scenarios such as deployment strategy, machine health checks, and retired node deletion. Exposing MachinePool Machines at the CAPI level will enable similar integrations with external projects. 
+
+For example, Cluster Autoscaler needs to delete individual Machines. Without a representation in the CAPI API of individual Machines in MachinePools, projects building integrations with Cluster API would need to be specialized to delete the cloud-specific resources for each MachinePool infrastructure provider. By exposing MachinePool Machines in CAPI, Cluster Autoscaler can treat Machines built by MachineDeployments and MachinePools in a uniform fashion through CAPI APIs.
+
+As another example, currently each MachinePool infrastructure provider must implement their own node drain functionality. When MachinePool Machines become available in the CAPI API, all providers can use its existing node drain implementation.
+
+### Goals
+
+- Enable CAPI MachinePool to create Machines to represent infrastructure provider MachinePool Machines
+- Ensure that Machines in MachinePools can participate in the same behaviors as MachineDeployment Machines (node drain, machine health checks, retired node deletion)
+- Enhance `clusterctl describe` to display MachinePool Machines
+- Allow CAPI MachinePool to include deployment strategies mirroring those of MachineDeployments (future work)
+
+### Non-Goals/Future Work
+
+- Require infrastructure providers to implement MachinePool Machines. This is intended to be an "opt-in" feature, since there are existing MachinePool implementations that cannot support it. For example, `AWSManagedMachinePool` does not allow for deleting individual instances.
```

Comment:
This is the part of the proposal that kind of concerns me because there will be two different types of representations and implementations for the machine pools (with and without machines), and this will be very confusing for users and system building on top of CAPI. 
Have we considered moving everything to a new representation (with machines)?

Reply:
I did consider for quite a while whether we should allow MachinePool Machines to be created even for `AKSManagedMachinePool`s and `AWSManagedMachinePool`s, as you suggested earlier. I can't convince myself that would be an improvement, because then the dichotomy is that deleting an MPM may or may not delete an actual resource.

So currently the proposal allows for a MachinePool implementation never to support MachinePool Machines, and that results in a difference in UX between an `AWSMachinePool` and an `AWSManagedMachinePool`, for example. I am not happy about that part, but I don't see a clearly better alternative.

Reply:
We decided that creating "CAPI-only" MachinePool Machines that can't be deleted by the user would be a more consistent UX, so I've updated the proposal to match that.

Reply:
> that can't be deleted by the user

Has the enhancement been updated to reflect this? I'd be interested to read how this is going to work and why this is a better UX than the alternate suggestions.